### ISSUE TITLE
parquet parsing breaks down after certain file size

### ISSUE THREAD

**Issue by v4lue4dded** [hyparam/hyparquet#2]
Unfortunately for me the parquet parsing breaks down after a certain file size.

I then get the following error:
```
encoding.js:258 Uncaught (in promise) RangeError: Maximum call stack size exceeded
    at readRleBitPackedHybrid (encoding.js:258:13)
    at readData (encoding.js:219:19)
    at readDefinitionLevels (datapage.js:153:55)
    at readDataPage (datapage.js:55:16)
    at readColumn (column.js:57:71)
    at read.js:152:24
    at async Promise.all (hyparquet/index 0)
    at async readRowGroup (read.js:209:3)
    at async parquetRead (read.js:53:25)
```

this is the python code that generated the files:
```python
# these files still work
df_measure_narrow_encoded.loc[:125000,].to_parquet(f"{export_dir}df_still_working_in_parquet_125000.parquet.txt", index = False)
df_measure_narrow_encoded.loc[:125000,].to_csv(f"{export_dir}df_still_working_in_parquet_125000.csv", index = False, encoding='utf-8-sig')

# these files don't work
df_measure_narrow_encoded.loc[:125125,].to_parquet(f"{export_dir}df_not_working_in_parquet_125125.parquet.txt", index = False)
df_measure_narrow_encoded.loc[:125125,].to_csv(f"{export_dir}df_not_working_in_parquet_125125.csv", index = False, encoding='utf-8-sig')
```

I'm providing example files both as parquet as well as csv files for completeness (the parquet files have a .txt ending since github wouldn't let me upload parquet). 

Here are the files that produce the described error
[df_not_working_in_parquet_125125.csv](https://github.com/hyparam/hyparquet/files/14895009/df_not_working_in_parquet_125125.csv)
[df_not_working_in_parquet_125125.parquet.txt](https://github.com/hyparam/hyparquet/files/14895016/df_not_working_in_parquet_125125.parquet.txt)

Here are the files that still work 
[df_still_working_in_parquet_125000.parquet.txt](https://github.com/hyparam/hyparquet/files/14895022/df_still_working_in_parquet_125000.parquet.txt)
[df_still_working_in_parquet_125000.csv](https://github.com/hyparam/hyparquet/files/14895020/df_still_working_in_parquet_125000.csv)

It seems to be a file size issue since it was the number of rows (not which rows, since I tested different row sets) that lead to the error.

You can easily recreate the error in my fork of your project (I added a complete file read and consol.log of the data based on your example and the commented out \\ TODO):
https://v4lue4dded.github.io/hyparquet/

Use the file selection import:
![image](https://github.com/hyparam/hyparquet/assets/61813337/bacc6484-f3bc-4a87-9a17-bebff8392b3e)

and you will get:
![image](https://github.com/hyparam/hyparquet/assets/61813337/ceaa7bd7-575d-46f0-8daf-cd13263d5662)

**Comment by platypii** [comment]
Thanks for the detailed example and steps to reproduce @v4lue4dded !

I found that I was able to load your file successfully in node.js but not in the browser. Further investigation showed that doing:

```js
rowData.push(...values)
```

Would throw "maximum call stack" error in the browser for large `values` arrays. Your example was 125k values.

Some experimentation showed that using `concat` is maybe less memory efficient, but avoids the max call stack error:

```js
rowData = rowData.concat(values)
```

I have updated hyparquet to use concat instead of spread operators. I just published v0.7.3 to npm. For me, this fixed your example file. But please try it and let me know. Thanks @v4lue4dded !!

**Comment by v4lue4dded** [comment]
Thanks for the fast reply platypii, I really appreciate it!!

Unfortunately the change that you implemented slowed parsing time of the package down by 10x or more for me :/ :
- the v0.7.2 of the package could load df_measure_narrow_encoded_tiny100000.parquet.txt in 260ms 
- the v0.7.3 of the package can load df_measure_narrow_encoded_tiny100000.parquet.txt in 2547ms

Furthermore the parsing time now slows down at an increasing rate as the number of rows increases:

| Number of rows | total parsing time | parsing time per 100k rows |
|----------------|--------------|----------------------------|
| 100k           | 2547 ms      | 2547 ms                    |
| 200k           | 10315 ms     | 5158 ms                    |
| 300k           | 22627 ms     | 7542 ms                    |
| 600k           | 98216 ms     | 16369 ms                   |
| 800k           | 201367 ms    | 25171 ms                   |

![image](https://github.com/hyparam/hyparquet/assets/61813337/b556f6ab-ddc5-47bb-90b5-8d4094549085)

I did this test using these github pages for version 0.7.3 and 0.7.2 respectively:
[https://v4lue4dded.github.io/hyparquet/](https://v4lue4dded.github.io/hyparquet/)
[https://v4lue4dded.github.io/hyparquet_old/](https://v4lue4dded.github.io/hyparquet_old/)

These are the files I used for the test:

[df_measure_narrow_encoded_tiny100000.parquet.txt](https://github.com/hyparam/hyparquet/files/14896981/df_measure_narrow_encoded_tiny100000.parquet.txt)
[df_measure_narrow_encoded_tiny200000.parquet.txt](https://github.com/hyparam/hyparquet/files/14896982/df_measure_narrow_encoded_tiny200000.parquet.txt)
[df_measure_narrow_encoded_tiny300000.parquet.txt](https://github.com/hyparam/hyparquet/files/14896983/df_measure_narrow_encoded_tiny300000.parquet.txt)
[df_measure_narrow_encoded_tiny600000.parquet.txt](https://github.com/hyparam/hyparquet/files/14896979/df_measure_narrow_encoded_tiny600000.parquet.txt)
[df_measure_narrow_encoded_tiny800000.parquet.txt](https://github.com/hyparam/hyparquet/files/14896977/df_measure_narrow_encoded_tiny800000.parquet.txt)

**Comment by platypii** [comment]
Yep I see the same performance regression as you. Oops!

I did some reading on stackoverflow, and it looks like there is no perfect solution between `push`, `concat`, and `for`. Each one is either slow, wastes memory, or crashes on chrome for large arrays. I came up with a helper function which tries to get the best of both worlds by chunking into 10000 value batches and using `push`:

```js
export function concat(aaa, bbb) {
  const chunk = 10000
  for (let i = 0; i < bbb.length; i += chunk) {
    aaa.push(...bbb.slice(i, i + chunk))
  }
}
```

I have published version 0.7.4 to npm. Please let me know if that works for you @v4lue4dded ! I care deeply about performance, so I really appreciate your testing and feedback!

**Comment by v4lue4dded** [comment]
With this version things load much faster, thanks! 😄
